### PR TITLE
bpo-40305: Fix server_close implementation for class ThreadingHTTPServer and TCPServer

### PR DESCRIPTION
**Pull request by RouxAntoine**
https://bugs.python.org/issue40305

* #84485

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@RouxAntoine

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/http/server.py [hunk 1]
```diff
@@ -144,6 +146,12 @@ def server_bind(self):
 class ThreadingHTTPServer(socketserver.ThreadingMixIn, HTTPServer):
     daemon_threads = True
 
+    def server_close(self):
+        # close TCP socket and more if override into HTTPServer class
+        HTTPServer.server_close(self)
+        # join all process request subthread
+        socketserver.ThreadingMixIn.server_close(self)
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: I do not understand why this change is needed. `ThreadingMixIn.server_close()` calls `super().server_close()`, which in this case should be `TCPServer.server_close()`, and `HTTPServer.server_close()` is the same as `TCPServer.server_close()`. Perhaps this change is outdated?

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Could you please add tests? It would help to understand what changes are really needed.

File: Lib/http/server.py (hunk 1 above)
Comment (RouxAntoine) [reply Lib/http/server.py]: Hello, that's an old subject, I found the code that led me to make this PR. I recall myself the subjet and come back to you with more explanation

Comment (RouxAntoine) [reply Lib/http/server.py]: ok I get it, you are right, the important part in this PR is the `super().shutdown()` in `TCPServer` (I.e. the change in 08e8c553c2b96eba3c40e41b7133e81f3b74ae97).

Imho this change (fcdcf39) in ThreadingHTTPServer is only for convenience. If someone extends `ThreadingHTTPServer` it should only call `super.server_close()` and which should delegate to both parent class. But I miss the things about mixin. So only `ThreadingMixIn.server_close(self)` is required if I properly understand isn't it ?

And in this case the newly `super.server_close()` will call `HTTPServer.server_close(self)` which is as you said `TCPServer.server_close()` and so with the seconds change (08e8c553c2b96eba3c40e41b7133e81f3b74ae97) will call `super().shutdown()` which is `BaseServer.shutdown(self)`


ps :  a toy project https://antoine-roux.tk/projects/python/bpo-40305 to reproduce and play with different case.

**Comment by serhiy-storchaka** [comment]
How hard is it to write some tests? Or at least a simple example that shows the difference?

**Comment by RouxAntoine** [comment]
> How hard is it to write some tests? Or at least a simple example that shows the difference?

ok it's good I successfully build and run test and adding a new test for this particular case.

File: Lib/http/server.py (hunk 1 above)
Comment (RouxAntoine) [reply Lib/http/server.py]: I try to add test for this code too, or deleting this code if actually useless

Comment (RouxAntoine) [reply Lib/http/server.py]: Ok right I remove this code it was indeed usefull

**Review by serhiy-storchaka — COMMENTED** [review]
Thank you for your test @RouxAntoine. Now I understand a little what the problem might be.

But I am not sure that calling `shutdown()` in `server_close()` is the right solution. `shutdown()` should only be called if `serve_forever()` was called, and necessarily in another thread. `serve_forever()` is not automatically called in any of these classes, it is called in your `WrappingThread`, therefore `shutdown()` should also be called in your code. Adding `self.server.shutdown()` before `self.server.server_close()` in `WrappingThread.stop()` solves the problem, isn't?

**Comment by RouxAntoine** [comment]
> Thank you for your test @RouxAntoine. Now I understand a little what the problem might be.
> 
> But I am not sure that calling `shutdown()` in `server_close()` is the right solution. `shutdown()` should only be called if `serve_forever()` was called, and necessarily in another thread. `serve_forever()` is not automatically called in any of these classes, it is called in your `WrappingThread`, therefore `shutdown()` should also be called in your code. Adding `self.server.shutdown()` before `self.server.server_close()` in `WrappingThread.stop()` solves the problem, isn't?

Yes I almost do that in my code year ago when a encounter the case. I just find this strange required call to `self.server.shutdown()` from parent method and `self.server.server_close()`. But feel free to close this pull request if you judge it should not be change.
